### DISCUSSION TITLE
Equality comparison of Headers to an object results in a TypeError

### DISCUSSION THREAD

**Discussion by bkrutov-pf** [encode/httpx#2421] [Potential Issue]
The following code, comparing a `Headers` instance to an object or anything that is not a `Headers`, `Mapping`, or iterable, results in a `TypeError` exception:
``` python
from httpx import Headers
Headers() == object()
```
```
Traceback (most recent call last):
  File "/usr/lib/python3.9/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "/home/borisk/.virtualenvs/python-sanic-extras/lib/python3.9/site-packages/httpx/_models.py", line 281, in __eq__
    other_headers = Headers(other)
  File "/home/borisk/.virtualenvs/python-sanic-extras/lib/python3.9/site-packages/httpx/_models.py", line 78, in __init__
    self._list = [
TypeError: 'object' object is not iterable
```

It looks like this code intends to avoid it, but apparently, `Headers.__init__` could raise a `TypeError` (and several other exceptions) in addition to a `ValueError` :
https://github.com/encode/httpx/blob/9e97d7d42963e408ef2949cc86515d6025dbec24/httpx/_models.py#L281-L285

**Comment by Kludex** [comment]
I guess `other` should be `object` instead of `Any`, and we should check for `TypeError` as well (?).

**Reply by Kludex** [reply]
What "several other exceptions"?

**Reply by bkrutov-pf** [reply]
By looking at the `Headers.__init__` code, I see that it could raise at least a `TypeError`, an `AttributeError` and a `UnicodeEncodeError` (both from `normalize_header_*`).
